File: static/assets/js/transact.js
// Function to update user name from session or fetch if not available
function updateUserName() {
    const userId = sessionStorage.getItem('user_id');
    const userName = sessionStorage.getItem('user_name');
    const userSurname = sessionStorage.getItem('user_surname');

    if (userId && userName && userSurname) {
        document.getElementById('user-name').textContent = `${userName} ${userSurname}`;
    } else {
        if (!userId) {
            console.error('User ID not found in session storage');
            return;
        }

        fetch(`/get_user_info?user_id=${userId}`)
            .then(response => response.json())
            .then(data => {
                if (data.error) {
                    console.error('Error:', data.error);
                } else {
                    document.getElementById('user-name').textContent = `${data.name_s} ${data.surname}`;
                    sessionStorage.setItem('user_name', data.name_s);
                    sessionStorage.setItem('user_surname', data.surname);
                }
            })
            .catch(error => console.error('Error fetching user info:', error));
    }
}

// Function to update account balances after transactions
function updateAccountBalances() {
    const userId = sessionStorage.getItem('user_id');

    fetch(`/get_accounts?user_id=${userId}`)
        .then(response => response.json())
        .then(data => {
            if (data.error) {
                console.error('Error:', data.error);
            } else {
                data.forEach(account => {
                    if (account.account_type === 'Debit') {
                        document.getElementById('debit-bal1').textContent = account.available_balance;
                        document.getElementById('debit-bal2').textContent = account.balance;
                    } else if (account.account_type === 'Savings') {
                        document.getElementById('savings-bal1').textContent = account.available_balance;
                        document.getElementById('savings-bal2').textContent = account.balance;
                    } else if (account.account_type === 'Credit') {
                        document.getElementById('credit-bal1').textContent = account.available_balance;
                        document.getElementById('credit-bal2').textContent = account.balance;
                    } else if (account.account_type === 'Investment') {
                        document.getElementById('invest-bal1').textContent = account.available_balance;
                        document.getElementById('invest-bal2').textContent = account.balance;
                    }
                });
            }
        })
        .catch(error => console.error('Error fetching account balances:', error));
}

// Event listeners for transaction buttons
document.addEventListener('DOMContentLoaded', function () {
    updateUserName();
    updateAccountBalances();
});

function handleTransfer() {
    const fromAccount = document.getElementById('transferDropdownFrom').value;
    const toAccount = document.getElementById('transferDropdownTo').value;
    const amount = parseFloat(document.getElementById('transferAmount').value);

    if (!fromAccount || !toAccount || isNaN(amount)) {
        alert('Please fill in all fields');
        return;
    }

    fetch('/transfer', {
        method: 'POST',
        headers: {
            'Content-Type': 'application/json'
        },
        body: JSON.stringify({ fromAccount, toAccount, amount })
    })
    .then(response => response.json())
    .then(data => {
        if (data.success) {
            alert('Transfer successful');
            location.reload();  // Reload the page on success
        } else {
            alert(data.message || 'Transfer failed');
        }
    })
    .catch(error => {
        alert('An error occurred while processing your request.');
    });
}

function handleWithdraw() {
    const fromAccount = document.getElementById('withdrawDropdownFrom').value;
    const amount = parseFloat(document.getElementById('withdrawAmount').value);

    if (!fromAccount || isNaN(amount)) {
        alert('Please fill in all fields');
        return;
    }

    fetch('/withdraw', {
        method: 'POST',
        headers: {
            'Content-Type': 'application/json'
        },
        body: JSON.stringify({ fromAccount, amount })
    })
    .then(response => response.json())
    .then(data => {
        if (data.success) {
            const code = data.withdrawalCode; // Assuming the response includes a withdrawal code
            alert(`Withdrawal successful. Your withdrawal code is: ${code}`);
            location.reload();  // Reload the page on success
        } else {
            alert(data.message || 'Withdrawal failed');
        }
    })
    .catch(error => {
        alert('An error occurred while processing your request.');
    });
}

File: transact.py
from flask import Flask, request, session, jsonify
from connect import get_db_connection
import random

app = Flask(__name__)
app.secret_key = '##################'

def transfer():
    data = request.get_json()

    if not data:
        return jsonify({'success': False, 'message': 'Invalid JSON input'})

    from_account = data.get('fromAccount')
    to_account = data.get('toAccount')
    amount = data.get('amount')

    if not from_account or not to_account or not amount:
        logging.error('Missing fields in input')
        return jsonify({'success': False, 'message': 'Please fill in all fields'})

    try:
        db = get_db_connection()
        cursor = db.cursor()
        cursor.execute("START TRANSACTION")
        cursor.execute("SELECT balance FROM Accounts WHERE account_type = %s AND user_id = %s FOR UPDATE", (from_account, session['user_id']))
        row = cursor.fetchone()

        if not row or row['balance'] < amount:
            return jsonify({'success': False, 'message': 'Insufficient funds or invalid account'})

        cursor.execute("UPDATE Accounts SET balance = balance - %s WHERE account_type = %s AND user_id = %s", (amount, from_account, session['user_id']))
        cursor.execute("UPDATE Accounts SET balance = balance + %s WHERE account_type = %s AND user_id = %s", (amount, to_account, session['user_id']))
        description = f"Transfer from {from_account} to {to_account}"
        cursor.execute("INSERT INTO Transactions (account_id, transaction_type, amount, description) VALUES ((SELECT account_id FROM Accounts WHERE account_type = %s AND user_id = %s), 'Cash Transfer', %s, %s)", (from_account, session['user_id'], amount, description))

        db.commit()
        return jsonify({'success': True})

    except Exception as e:
        if db:
            db.rollback()
        return jsonify({'success': False, 'message': str(e)})

    finally:
        if cursor:
            cursor.close()
        if db:
            db.close()

def withdraw():
    data = request.get_json()

    if not data:
        return jsonify({'success': False, 'message': 'Invalid JSON input'})

    from_account = data.get('fromAccount')
    amount = data.get('amount')

    if not from_account or not amount:
        return jsonify({'success': False, 'message': 'Please fill in all fields'})

    try:
        db = get_db_connection()
        cursor = db.cursor()
        cursor.execute("START TRANSACTION")

        cursor.execute("SELECT balance FROM Accounts WHERE account_type = %s AND user_id = %s FOR UPDATE", (from_account, session['user_id']))
        row = cursor.fetchone()

        if not row or row['balance'] < amount:
            return jsonify({'success': False, 'message': 'Insufficient funds or invalid account'})

        cursor.execute("UPDATE Accounts SET balance = balance - %s WHERE account_type = %s AND user_id = %s", (amount, from_account, session['user_id']))

        description = "Cash Withdrawal"
        cursor.execute("INSERT INTO Transactions (account_id, transaction_type, amount, description) VALUES ((SELECT account_id FROM Accounts WHERE account_type = %s AND user_id = %s), 'Withdrawal', %s, %s)", (from_account, session['user_id'], amount, description))

        code = f'XEXWITHDR-{random.randint(100000, 999999)}'
        user_email = fetch_user_email(session['user_id'])
        if user_email:
            code = f'XEXWITHDR-{random.randint(100000, 999999)}'

        db.commit()
        logging.debug("Transaction committed")
        return jsonify({'success': True, 'withdrawalCode': code})

    except Exception as e:
        if db:
            db.rollback()
        logging.error(f"Exception: {str(e)}")
        return jsonify({'success': False, 'message': str(e)})

    finally:
        if cursor:
            cursor.close()
        if db:
            db.close()

def fetch_user_email(user_id):

    try:
        db = get_db_connection()
        cursor = db.cursor()

        cursor.execute("SELECT email FROM Users WHERE user_id = %s", (user_id,))
        row = cursor.fetchone()
        if row:
            return row[0]
        return None

    except Exception as e:
        return None

    finally:
        if cursor:
            cursor.close()
        if db:
            db.close()
